### PR TITLE
SNS governance, show neuron reward portion of each reward-event.

### PR DESCRIPTION
**Pull request by levifeldman**

### CODE REVIEW

**Review by aterga — COMMENTED** [review]
Thanks for your PR. I'll get back to you when I find out the exact process for external contributions to the SNS. In the meantime, here's some feedback on your code.

PS. Please note that this feature would need unit tests in order to be accepted.

**Comment by levifeldman** [comment]
Awesome, thanks for your help and feedback.

File: rs/sns/governance/src/governance.rs [hunk 1]
```diff
@@ -4935,6 +4938,18 @@ impl Governance {
...
+                // Insert the neuron_reward_e8s into the reward_events_to_neuron_reward_e8s map on the neuron. 
+                neuron.reward_events_to_neuron_reward_e8s.insert( 
+                    reward_event_end_timestamp_seconds, 
+                    neuron_reward_e8s
+                );
+                // prune
+                while neuron.reward_events_to_neuron_reward_e8s.len() > MAX_KEEP_REWARD_EVENTS_TO_NEURON_REWARD_E8S_PER_NEURON as usize {
+                    // prune earliest.
+                    neuron.reward_events_to_neuron_reward_e8s.remove(
+                        &neuron.reward_events_to_neuron_reward_e8s.keys().min().unwrap().clone()
```
[the remark this answers is not in the document]
Comment (levifeldman) [reply rs/sns/governance/src/governance.rs:4950]: It does at first glance but Rust's borrow checker makes sure that a type cannot be borrowed mutably while an immutable-borrow exists. In this case the .keys() iterator returns immutable-references to the map's keys and the map.remove method borrows self mutably so map.remove cannot be called if there exists any immutable-references to the map so it needs to be cloned to drop the immutable-reference. 

I tried without the clone here is the error message:
image

File: rs/sns/governance/proto/ic_sns_governance/pb/v1/governance.proto [hunk 2]
```diff
@@ -211,6 +211,11 @@ message Neuron {
   // with the oldest entries first, i.e. it holds for all i that:
   // entry[i].timestamp_of_disbursement_seconds <= entry[i+1].timestamp_of_disbursement_seconds
   repeated DisburseMaturityInProgress disburse_maturity_in_progress = 18;
+
+  // The keys are the reward-event-timestamps (end_timestamp_seconds) and the values are the neuron_reward_e8s for this reward-event.   
+  // Stores minimum 5 latest reward events.
+  map<uint64, uint64> reward_events_to_neuron_reward_e8s = 19;
```
[the remark this answers is not in the document]
Comment (levifeldman) [reply rs/sns/governance/proto/ic_sns_governance/pb/v1/governance.proto:217]: :+1:

File: rs/sns/governance/proto/ic_sns_governance/pb/v1/governance.proto [hunk 3]
```diff
@@ -211,6 +211,11 @@ message Neuron {
   // with the oldest entries first, i.e. it holds for all i that:
   // entry[i].timestamp_of_disbursement_seconds <= entry[i+1].timestamp_of_disbursement_seconds
   repeated DisburseMaturityInProgress disburse_maturity_in_progress = 18;
+
+  // The keys are the reward-event-timestamps (end_timestamp_seconds) and the values are the neuron_reward_e8s for this reward-event.   
+  // Stores minimum 5 latest reward events.
```
[the remark this answers is not in the document]
Comment (levifeldman) [reply rs/sns/governance/proto/ic_sns_governance/pb/v1/governance.proto:216]: :+1:

File: rs/sns/governance/src/governance.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (levifeldman) [reply rs/sns/governance/src/governance.rs:4950]: I changed it to use map.pop_first().

File: rs/sns/governance/src/governance.rs [hunk 4]
```diff
@@ -4935,6 +4938,18 @@ impl Governance {
                 } else {
                     neuron.maturity_e8s_equivalent += neuron_reward_e8s;
                 }
+                // Insert the neuron_reward_e8s into the reward_events_to_neuron_reward_e8s map on the neuron. 
+                neuron.reward_events_to_neuron_reward_e8s.insert( 
+                    reward_event_end_timestamp_seconds, 
+                    neuron_reward_e8s
+                );
+                // prune
+                while neuron.reward_events_to_neuron_reward_e8s.len() > MAX_KEEP_REWARD_EVENTS_TO_NEURON_REWARD_E8S_PER_NEURON as usize {
```
[the remark this answers is not in the document]
Comment (levifeldman) [reply rs/sns/governance/src/governance.rs:4947]: :+1: I wasn't sure if I could rely on the protobuf-to-rust conversion always generating a BTreeMap but now using map.pop_first is only possible on a BTreeMap so if the protobuf-to-rust conversion ever changes to a HashMap then it will be noticed and handled. 

It is cool how you choose to prune the map before insertion, good to get that perspective.

**Comment by levifeldman** [comment]
Hey @aterga, I pushed changes to this branch on my fork and can be viewed here: https://github.com/levifeldman/ic/compare/master...levifeldman-show-neuron-reward-portion.

**Comment by levifeldman** [comment]
Hey @aterga, I pushed changes to this branch on my fork and can be viewed here: https://github.com/levifeldman/ic/compare/master...levifeldman-show-neuron-reward-portion.

File: rs/sns/governance/src/neuron.rs [hunk 5]
```diff
@@ -24,6 +24,9 @@ pub const MAX_LIST_NEURONS_RESULTS: u32 = 100;
 /// The default voting_power_percentage_multiplier applied to a neuron.
 pub const DEFAULT_VOTING_POWER_PERCENTAGE_MULTIPLIER: u64 = 100;
 
+// The max number of reward_events_to_neuron_reward_e8s per neuron
+pub const MAX_KEEP_REWARD_EVENTS_TO_NEURON_REWARD_E8S_PER_NEURON: u32 = 5;
```
[the remark this answers is not in the document]
Comment (levifeldman) [reply rs/sns/governance/src/neuron.rs:28]: If it's set to 1 the caller would need to sync at least 2/3/4 times per day to make sure to get the data. If a caller only syncs once per day then it might miss the reward-event because a sns-reward-event can happen in between the different list_neurons calls and the caller will not know if the neurons (in the list_neurons chunks after the sns-reward-event) got their reward for the previous reward-event or if they were not eligible for rewards since the sns-reward-event just pruned them. So since here it is > 1, the caller canister can call once a day, and start from the second-to-earliest reward-event in the map and that way the reward-event will for sure be there on all neurons that voted for that second-to-earliest reward-event even if a sns-reward-event happens in between the list_neurons calls as long as the list_neurons calls take less than a day to go through all the neurons. 

Another reason is if a caller canister needs maintenance for a day its better if it can catch up. I think one day window to sync the data is too cramped.

Combining these reasons gives a value of 3 for the constant, and then add 2 for good measure to make 5. That's how I think about it. I am cool with setting the value of the constant to 3, if you think that is better.

**Comment by levifeldman** [comment]
I agreed to the cla for contributing to DFINITY's repos and the record for that is here: https://github.com/dfinity/cla/issues/15.

**Comment by levifeldman** [comment]
I pushed a unit-test, let me know if there is a corner case you want me to add into it. Now it checks to make sure the neuron maps get inserted with the neuron_reward_e8s. One thing I will add Wednesday is a check to make sure the map gets pruned. Now it checks

**Comment by levifeldman** [comment]
The unit-test now checks that the earliest reward event gets pruned when needed and that the map length stays <= the max-keys.
